

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>hps.core package &mdash; HPS-Urry  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="hps.datasets package" href="hps.datasets.html" />
    <link rel="prev" title="hps package" href="hps.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> HPS-Urry
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">hps_urry</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="hps.html">hps package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="hps.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">hps.core package</a></li>
<li class="toctree-l4"><a class="reference internal" href="hps.datasets.html">hps.datasets package</a></li>
<li class="toctree-l4"><a class="reference internal" href="hps.parameters.html">hps.parameters package</a></li>
<li class="toctree-l4"><a class="reference internal" href="hps.reporter.html">hps.reporter package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="hps.html#module-hps">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">HPS-Urry</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="modules.html">hps_urry</a> &raquo;</li>
        
          <li><a href="hps.html">hps package</a> &raquo;</li>
        
      <li>hps.core package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/hps.core.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="hps-core-package">
<h1>hps.core package<a class="headerlink" href="#hps-core-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-hps.core.geometry">
<span id="hps-core-geometry-module"></span><h2>hps.core.geometry module<a class="headerlink" href="#module-hps.core.geometry" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="hps.core.geometry.geometry">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">hps.core.geometry.</span></span><span class="sig-name descname"><span class="pre">geometry</span></span><a class="headerlink" href="#hps.core.geometry.geometry" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A class to hold methods for calculating geometrical values 
given sets of atom coordinates.</p>
<dl class="py method">
<dt class="sig sig-object py" id="hps.core.geometry.geometry.angle">
<span class="sig-name descname"><span class="pre">angle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coord2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coord3</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.geometry.geometry.angle" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the angle length between three (x,y,z) quantity coordinates.</p>
<dl class="simple">
<dt>coord1<span class="classifier">simtk.unit.quantity.Quantity array</span></dt><dd><p>Vector for the first coordinate.</p>
</dd>
<dt>coord2<span class="classifier">simtk.unit.quantity.Quantity array</span></dt><dd><p>Vector for the second coordinate.</p>
</dd>
<dt>coord3<span class="classifier">simtk.unit.quantity.Quantity array</span></dt><dd><p>Vector for the third coordinate.</p>
</dd>
</dl>
<dl class="simple">
<dt>simtk.unit.quantity.Quantity</dt><dd><p>Quantity (value and unit) of the angle length in radians.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.geometry.geometry.bond">
<span class="sig-name descname"><span class="pre">bond</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coord2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.geometry.geometry.bond" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the distance length between two (x,y,z) quantity coordinates.</p>
<dl class="simple">
<dt>coord1<span class="classifier">simtk.unit.quantity.Quantity array</span></dt><dd><p>Vector for the first coordinate.</p>
</dd>
<dt>coord2<span class="classifier">simtk.unit.quantity.Quantity array</span></dt><dd><p>Vector for the second coordinate.</p>
</dd>
</dl>
<dl class="simple">
<dt>simtk.unit.quantity.Quantity</dt><dd><p>Quantity (value and unit) of the distance length in nanometers.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.geometry.geometry.position2Array">
<span class="sig-name descname"><span class="pre">position2Array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_unit</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.geometry.geometry.position2Array" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts an OpenMM position object quantity into a numpy array.</p>
<dl class="simple">
<dt>position<span class="classifier">simtk.unit.quantity.Quantity</span></dt><dd><p>Array containing quantity objects [e.g. (x,y,z) array returned
from positions].</p>
</dd>
<dt>output_unit<span class="classifier">simtk.unit.unit.Unit</span></dt><dd><p>Unit in which to return the items of the array.</p>
</dd>
</dl>
<dl class="simple">
<dt>numpy.ndarray</dt><dd><p>A numpy array containing the quantity values converted to floats.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.geometry.geometry.torsion">
<span class="sig-name descname"><span class="pre">torsion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coord2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coord3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coord4</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.geometry.geometry.torsion" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the torsion angle length between four (x,y,z) quantity 
coordinates.</p>
<dl class="simple">
<dt>coord1<span class="classifier">simtk.unit.quantity.Quantity array</span></dt><dd><p>Vector for the first coordinate.</p>
</dd>
<dt>coord2<span class="classifier">simtk.unit.quantity.Quantity array</span></dt><dd><p>Vector for the second coordinate.</p>
</dd>
<dt>coord3<span class="classifier">simtk.unit.quantity.Quantity array</span></dt><dd><p>Vector for the third coordinate.</p>
</dd>
<dt>coord4<span class="classifier">simtk.unit.quantity.Quantity array</span></dt><dd><p>Vector for the fourth coordinate.</p>
</dd>
</dl>
<dl class="simple">
<dt>simtk.unit.quantity.Quantity</dt><dd><p>Quantity (value and unit) of the torsion length in radians.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-hps.core.models">
<span id="hps-core-models-module"></span><h2>hps.core.models module<a class="headerlink" href="#module-hps.core.models" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="hps.core.models.models">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">hps.core.models.</span></span><span class="sig-name descname"><span class="pre">models</span></span><a class="headerlink" href="#hps.core.models.models" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A class to hold functions for the automated generation of default SBM models.</p>
<dl class="simple">
<dt>getCAModel(structure_file, kwarg**)</dt><dd><p>Creates an alpha-carbon only sbmOpenMM system class object with default
initialized parameters.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="hps.core.models.models.getCAModel">
<span class="sig-name descname"><span class="pre">getCAModel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">create_system</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residue_masses</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residue_radii</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residue_charge</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residue_hps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forcefield_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.models.models.getCAModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialises a coarse-grained, carbon alpha (CA), sbmOpenMM system class
from a structure and a contact file defining the native contacts for the
coarse grained model.</p>
<p>The system creation steps are:</p>
<ol class="arabic simple">
<li><p>Add the geometrical parameters for the model.</p></li>
<li><p>Add the default force field parameters for the model.</p></li>
<li><p>Create the default force objects.</p></li>
<li><p>Create the OpenMM system class.</p></li>
</ol>
<p>The method can be used to generate an initialized sbmOpenMM system class, that only
contains the geometrical parameters, by passing the option default_parameters as False.
This is useful to store the geometrical values of bonds, angles, dihedrals, etc. in
order to add custom parameters and forces.</p>
<p>The method can also be created without the initialisation of the forces classes by
setting default_forces to False. This allows to load the default forcefield parameters
and to modified them before creating the OpenMM force objects.</p>
<p>The method can also be stopped before creating the OpenMM system class using create_system=False.</p>
<p>Finally, a forcefield file can be given in order to read the forcefield parameters from it. You can give
None to the contact file argument if you provide a forcefield file. Contacts definitions will be extracted from this
file even if you give a path to a contact file.</p>
<dl class="simple">
<dt>structure_file<span class="classifier">string</span></dt><dd><p>Path to the input structure file.</p>
</dd>
<dt>default_parameters<span class="classifier">boolean (True)</span></dt><dd><p>Whether to add default SBM CA forcefield parameters to the model.</p>
</dd>
<dt>default_forces<span class="classifier">boolean (True)</span></dt><dd><p>Whether to initialize default SBM CA force objects. Set to False if
the parameters will be different from the default ones.</p>
</dd>
<dt>create_system<span class="classifier">boolean (True)</span></dt><dd><p>If True the function will call the createSystemObject() method
to create an OpenMM system object. If modifications to the default
forcefield are necessary this option should be given False.</p>
</dd>
<dt>minimize<span class="classifier">boolean (False)</span></dt><dd><p>If True the initial structure will undergo the energy minimization.</p>
</dd>
<dt>residue_masses<span class="classifier">boolean (True)</span></dt><dd><p>Set each alpha carbon atom mass to its average amino acid residue mass.</p>
</dd>
<dt>residue_radii<span class="classifier">boolean (True)</span></dt><dd><p>Set each alpha carbon atom radius to its statistical amino acid residue radius.</p>
</dd>
<dt>residue_charge<span class="classifier">boolean (True)</span></dt><dd><p>set charge for atoms in the system</p>
</dd>
<dt>residue_hps<span class="classifier">boolean (True)</span></dt><dd><p>set HPS scale for atoms in the system</p>
</dd>
<dt>forcefield_file<span class="classifier">string</span></dt><dd><p>Path to the input forcefield file.</p>
</dd>
</dl>
<dl class="simple">
<dt>sbm<span class="classifier">sbmOpenMM.system</span></dt><dd><p>Initialized sbmOpenMM.system class with default options for defining
a coarse-grained CA SBM force field.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-hps.core.system">
<span id="hps-core-system-module"></span><h2>hps.core.system module<a class="headerlink" href="#module-hps.core.system" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="hps.core.system.system">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">hps.core.system.</span></span><span class="sig-name descname"><span class="pre">system</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">particles_mass</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A class containing methods and parameters for generating Structure Based
Models (SBM) systems to be simulated using the OpenMM interface. It offers
flexibility to create default and custom SBM systems and to easily
modify their parameters.</p>
<dl class="simple">
<dt>structure_path<span class="classifier">string</span></dt><dd><p>Path to the pdb or cif input file.</p>
</dd>
<dt>structure<span class="classifier">openmm.app.pdbfile.PDBFile or openmm.app.pdbxfile.PDBxFile</span></dt><dd><p>Object that holds the information of OpenMM PDB or CIF parsing methods.</p>
</dd>
<dt>topology<span class="classifier">openmm.app.topology.Topology</span></dt><dd><p>OpenMM topology of the model.</p>
</dd>
<dt>positions<span class="classifier">unit.quantity.Quantity</span></dt><dd><p>Atomic positions of the model.</p>
</dd>
<dt>particles_mass<span class="classifier">float or list</span></dt><dd><p>Mass of each particle. If float then uniform masses are given to all
particles. If list per-particle masses are assigned.</p>
</dd>
<dt>model_type<span class="classifier">string</span></dt><dd><p>String representing the model type: All-atom (AA), alpha-carbon (CA)
or multi-basin variants (AA-MB, CA-MB).</p>
</dd>
<dt>atoms<span class="classifier">list</span></dt><dd><p>A list of the current atoms in the model. The items are simtk.openmm.app.topology.Atom
initialised classes.</p>
</dd>
<dt>n_atoms<span class="classifier">int</span></dt><dd><p>Total numer of atoms in the model.</p>
</dd>
<dt>bonds<span class="classifier">collections.OrderedDict</span></dt><dd><p>A dict that uses bonds (2-tuple of simtk.openmm.app.topology.Atom objects)
present in the model as keys and their forcefield properties as values.</p>
</dd>
<dt>bonds_indexes<span class="classifier">list</span></dt><dd><p>A list containing the zero-based indexes of the atoms defining the bonds in the model.</p>
</dd>
<dt>n_bonds<span class="classifier">int</span></dt><dd><p>Total numer of bonds in the model.</p>
</dd>
<dt>harmonicBondForce<span class="classifier">openmm.HarmonicBondForce</span></dt><dd><p>Stores the OpenMM HarmonicBondForce initialised-class. Implements
an harmonic bond potential between pairs of particles, that depends
quadratically on their distance.</p>
</dd>
<dt>yukawaForce<span class="classifier">openmm.CustomNonbondedForce</span></dt><dd><p>Stores the OpenMM CustomNonbondedForce initialized-class.
Implements the Debye-Huckle potential.</p>
</dd>
<dt>pairWiseForce<span class="classifier">openmm.CustomNonbondedForce</span></dt><dd><p>Stores the OpenMM CustomNonbondedForce initialized-class. Implements the pairwise short-range
potential.</p>
</dd>
<dt>forceGroups<span class="classifier">collections.OrderedDict</span></dt><dd><p>A dict that uses force names as keys and their corresponding force
as values.</p>
</dd>
<dt>system<span class="classifier">openmm.System</span></dt><dd><p>Stores the OpenMM System initialised class. It stores all the forcefield
information for the SBM model.</p>
</dd>
<dt>rf_sigma<span class="classifier">float</span></dt><dd><p>Sigma parameter used in the pairwise force object.</p>
</dd>
<dt>rf_cutoff<span class="classifier">float</span></dt><dd><p>Cutoff value used for repulsion force interactions.</p>
</dd>
<dt>exclusion_NB<span class="classifier">NxN matrix</span></dt><dd><p>This is a pairwise interaction matrix.
initialize by np.ones((N, N)), N is the number of beads in the system.
By default, all atoms interact via pairwise. but this is not true, we should
exclude atom pairs that covalently bonded by when adding Harmonic bonds, we set matrix element of (i,j), (j,i)
to 0, where i and j are atom indices in bond.</p>
</dd>
</dl>
<dl class="simple">
<dt>removeHydrogens()</dt><dd><p>Remove hydrogens from the input structure by using a regexpression pattern.
Used specially for creating all atom (AA) models.</p>
</dd>
<dt>getCAlphaOnly()</dt><dd><p>Filter in only alpha carbon atoms from the input structure and updates
the topology object to add new bonds between them. Used specially for
creating alpha-carbon (CA) corse-grained models.</p>
</dd>
<dt>getAtoms()</dt><dd><p>Reads atoms from topology, adds them to the main class and sorts them
into a dictionary to store their forcefield properties.</p>
</dd>
<dt>getBonds()</dt><dd><p>Reads bonds from topology, adds them to the main class and sorts them
into a dictionary to store their forcefield properties.</p>
</dd>
<dt>setBondParameters()</dt><dd><p>Change the forcefield parameters for bonded terms.</p>
</dd>
<dt>setParticlesMasses()</dt><dd><p>Change the mass parameter for each atom in the system.</p>
</dd>
<dt>setParticlesRadii()</dt><dd><p>Change the excluded volume radius parameter for each atom in the system.</p>
</dd>
<dt>addHarmonicBondForces()</dt><dd><p>Creates an harmonic bonded force term for each bond in the main
class using their defined forcefield parameters.</p>
</dd>
<dt>addYukawaForces()</dt><dd><p>Creates a nonbonded force term for electrostatic interaction DH potential.</p>
</dd>
<dt>addPairWiseForces()</dt><dd><p>Creates a nonbonded force term for pairwise interaction (customize LJ 12-6 potential).</p>
</dd>
<dt>createSystemObject()</dt><dd><p>Creates OpenMM system object adding particles, masses and forces.
It also groups the added forces into Force-Groups for the sbmReporter
class.</p>
</dd>
<dt>addParticles()</dt><dd><p>Add particles to the system OpenMM class instance.</p>
</dd>
<dt>addSystemForces()</dt><dd><p>Add forces to the system OpenMM class instance. It also save
names for the added forces to include them in the reporter class.</p>
</dd>
<dt>dumpStructure()</dt><dd><p>Writes a structure file of the system in its current state.</p>
</dd>
<dt>dumpTopology()</dt><dd><p>Writes a topology file of the system in PSF format, this is used for visualization.
Be careful when using this for analysis since the atom masses are not correct,
it specifies every masses equal to Carbon masses.</p>
</dd>
<dt>dumpForceFieldData()</dt><dd><p>Writes to a file the parameters of the SBM forcefield.</p>
</dd>
<dt>loadForcefieldFromFile()</dt><dd><p>Loads forcefield parameters from a sbmOpenMM force field file written with
the dumpForceFieldData() method.</p>
</dd>
<dt>setCAMassPerResidueType()</dt><dd><p>Sets alpha carbon atoms to their average residue mass. Used specially for
modifying alpha-carbon (CA) coarse-grained models.</p>
</dd>
<dt>setCARadiusPerResidueType()</dt><dd><p>Sets alpha carbon atoms to their average residue mass. Used specially for
modifying alpha-carbon (CA) coarse-grained models.</p>
</dd>
<dt>setCAHPSPerResidueType()</dt><dd><p>Sets alpha carbon atoms to their residue Urry Hydropathy scale. Used specially for
modifying alpha-carbon (CA) coarse-grained models.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.addHarmonicBondForces">
<span class="sig-name descname"><span class="pre">addHarmonicBondForces</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.addHarmonicBondForces" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates an openmm.HarmonicBondForce() object with the bonds and
parameters set up in the “bonds” dictionary attribute. The force object
is stored at the “harmonicBondForce” attribute.</p>
<p>The force parameters must be contained in self.bonds as follows:
self.bonds is a dictionary:</p>
<blockquote>
<div><ul>
<li><p>The keys are 2-tuples for two atom items in self.topology.atoms attribute.</p></li>
<li><p>The values are a 2-tuple of parameters in the following order:</p>
<blockquote>
<div><ul class="simple">
<li><p>first  -&gt; bond0 (quantity)</p></li>
<li><p>second -&gt; k (float)</p></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
<p>None</p>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.addPairWiseForces">
<span class="sig-name descname"><span class="pre">addPairWiseForces</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.addPairWiseForces" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates an openmm.CustomNonbondedForce() object with the parameters
sigma and epsilon given to this method. The custom non-bonded force
is initialized with the formula: (note: hps here is lambda0 in the paper)</p>
<p>Unlike BondForce class, where we specify index for atoms pair to add bond, it means
that number of bondForces may differ from number of particle.
NonBondedForce is add to all particles, hence we don’t need pass the atom index.</p>
<dl>
<dt>energy = ‘</dt><dd><p>step(2^(1/6)*sigma - r) * (4*epsilon* ((sigma/r)^12-(sigma/r)^6) + (1-muy*hps+delta)*epsilon )
+
(1-step(2^(1/6)-r)) * ((muy*hps-delta)*4*epsilon*((sigma/r)^12-(sigma/r)^6))</p>
<p>sigma   = 0.5*(sigma1+sigma2)
hps  = 0.5*(hps1+hps2)
muy     = 1
delta   = 0.08
epsilon = 0.2 kcal/mol</p>
</dd>
</dl>
<p>‘
The force object is stored at the “pairWiseForce” attribute.</p>
<dl class="simple">
<dt>epsilon<span class="classifier">float</span></dt><dd><p>Value of the epsilon constant in the energy function.</p>
</dd>
<dt>sigma<span class="classifier">float or list</span></dt><dd><p>Value of the sigma constant (in nm) in the energy function. If float the
same sigma value is used for every particle. If list a unique
parameter is given for each particle.</p>
</dd>
<dt>cutoff<span class="classifier">float</span></dt><dd><p>The cutoff distance (in nm) being used for the nonbonded interactions.</p>
</dd>
</dl>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.addParticles">
<span class="sig-name descname"><span class="pre">addParticles</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.addParticles" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a particle to the system for each atom in it. The mass
of each particle is set up with the values in the ‘particles_mass’
attribute.</p>
<p>None</p>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.addSystemForces">
<span class="sig-name descname"><span class="pre">addSystemForces</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.addSystemForces" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds generated forces to the system, also adding
a force group to the ‘forceGroups’ attribute dictionary.</p>
<p>None</p>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.addYukawaForces">
<span class="sig-name descname"><span class="pre">addYukawaForces</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.addYukawaForces" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates an openmm.CustomNonbondedForce() object with the parameters
sigma and epsilon given to this method. The custom non-bonded force
is initialized with the formula:</p>
<p>energy = ‘factor*charge1*charge2/epsilon/r*exp(-r/lD)’
where factor is for short to convert dimensionless in calculation to kj.nm/mol/e^2 unit</p>
<p>The force object is stored at the “yukawaForce” attribute.</p>
<p>None</p>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.checkBondDistances">
<span class="sig-name descname"><span class="pre">checkBondDistances</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.39</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.checkBondDistances" title="Permalink to this definition">¶</a></dt>
<dd><p>Searches for large bond distances for the atom pairs defined in
the ‘bonds’ attribute. It raises an error when large bonds are found.</p>
<dl class="simple">
<dt>threshold<span class="classifier">float</span></dt><dd><p>Threshold to check for large bond distances.</p>
</dd>
</dl>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.checkLargeForces">
<span class="sig-name descname"><span class="pre">checkLargeForces</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.checkLargeForces" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints the SBM system energies of the input configuration of the
system. It optionally checks for large forces acting upon all
particles in the SBM system and iteratively minimizes the system
configuration until no forces larger than a threshold are found.</p>
<dl class="simple">
<dt>threshold<span class="classifier">float</span></dt><dd><p>Treshold to check for large forces.</p>
</dd>
<dt>minimize<span class="classifier">float</span></dt><dd><p>Whether to iteratively minimize the system until all forces are lower or equal to
the thresshold value.</p>
</dd>
</dl>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.createSystemObject">
<span class="sig-name descname"><span class="pre">createSystemObject</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">check_bond_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_large_forces</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bond_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.39</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.createSystemObject" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates an openmm.System() object using the force field parameters
given to the SBM ‘system’ class. It adds particles, forces and
creates a force group for each force object. Optionally the method
can check for large bond distances (default) and minimize the atomic
positions if large forces are found in any atom (default False).</p>
<dl class="simple">
<dt>minimize<span class="classifier">boolean (False)</span></dt><dd><p>Whether to minimize the system if large forces are found.</p>
</dd>
<dt>check_bond_distances<span class="classifier">boolean (True)</span></dt><dd><p>Whether to check for large bond distances.</p>
</dd>
<dt>check_large_forces<span class="classifier">boolean (False)</span></dt><dd><p>Whether to print force summary of force groups</p>
</dd>
<dt>force_threshold<span class="classifier">float (10.0)</span></dt><dd><p>Treshold to check for large forces.</p>
</dd>
<dt>bond_threshold<span class="classifier">float (0.39)</span></dt><dd><p>Treshold to check for large bond distances.</p>
</dd>
</dl>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.dumpForceFieldData">
<span class="sig-name descname"><span class="pre">dumpForceFieldData</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.dumpForceFieldData" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes a file containing the current forcefield parameters in the
sbmOpenMM system.</p>
<dl class="simple">
<dt>output_file<span class="classifier">string</span></dt><dd><p>name of the output file.</p>
</dd>
</dl>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.dumpStructure">
<span class="sig-name descname"><span class="pre">dumpStructure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.dumpStructure" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes a PDB file containing the currently defined abmOpenMM system atoms.</p>
<dl class="simple">
<dt>output_file<span class="classifier">string</span></dt><dd><p>name of the PDB output file.</p>
</dd>
</dl>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.dumpTopology">
<span class="sig-name descname"><span class="pre">dumpTopology</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.dumpTopology" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes a file containing the current topology in the
sbmOpenMM system.</p>
<dl class="simple">
<dt>output_file<span class="classifier">string</span></dt><dd><p>name of the output file.</p>
</dd>
</dl>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.getAtoms">
<span class="sig-name descname"><span class="pre">getAtoms</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.getAtoms" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds atoms in the OpenMM topology instance to the sbmOpenMM system class.</p>
<p>None</p>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.getBonds">
<span class="sig-name descname"><span class="pre">getBonds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">except_chains</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.getBonds" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds bonds in the OpenMM topology instance to the sbmOpenMM system class.</p>
<p>None</p>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.getCAlphaOnly">
<span class="sig-name descname"><span class="pre">getCAlphaOnly</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.getCAlphaOnly" title="Permalink to this definition">¶</a></dt>
<dd><p>Keeps in the SBM system only the alpha carbon atoms from the OpenMM topology.</p>
<p>None</p>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.loadForcefieldFromFile">
<span class="sig-name descname"><span class="pre">loadForcefieldFromFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">forcefield_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.loadForcefieldFromFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads force field parameters from a force field file written by the
dumpForceFieldData() method into the sbmOpenMM system.
NOTE: I still keep this function since I think I will use it later.
I have not customized this function yet so do not use loadForcefieldFromFile function in simulation.
Just keep here for prototype and customize later…</p>
<dl class="simple">
<dt>forcefield_file<span class="classifier">string</span></dt><dd><p>path to the force field file.</p>
</dd>
</dl>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.removeHydrogens">
<span class="sig-name descname"><span class="pre">removeHydrogens</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">except_chains</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.removeHydrogens" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes all hydrogen atoms in the topology from the SBM system.</p>
<p>None</p>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.setBondParameters">
<span class="sig-name descname"><span class="pre">setBondParameters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bond_parameters</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.setBondParameters" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the harmonic bond constant force parameters. The input can be
a float, to set the same parameter for all force interactions, or
a list, to define a unique parameter for each force interaction.</p>
<dl class="simple">
<dt>bond_parameters<span class="classifier">float or list</span></dt><dd><p>Parameter(s) to set up for the harmonic bond forces.</p>
</dd>
</dl>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.setCAChargePerResidueType">
<span class="sig-name descname"><span class="pre">setCAChargePerResidueType</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.setCAChargePerResidueType" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the charge of the alpha carbon atoms
to characteristic charge of their corresponding amino acid
residue.</p>
<p>None</p>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.setCAHPSPerResidueType">
<span class="sig-name descname"><span class="pre">setCAHPSPerResidueType</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.setCAHPSPerResidueType" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the HPS of the alpha carbon atoms.
The current implementation is using Urry scale.</p>
<p>None</p>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.setCAMassPerResidueType">
<span class="sig-name descname"><span class="pre">setCAMassPerResidueType</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.setCAMassPerResidueType" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the masses of the alpha carbon atoms to the average mass
of its amino acid residue.</p>
<p>None</p>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.setCARadiusPerResidueType">
<span class="sig-name descname"><span class="pre">setCARadiusPerResidueType</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.setCARadiusPerResidueType" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the excluded volume radii of the alpha carbon atoms
to characteristic radii of their corresponding amino acid
residue.</p>
<p>None</p>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.setParticlesCharge">
<span class="sig-name descname"><span class="pre">setParticlesCharge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">particles_charge</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.setParticlesCharge" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the charge of the particles in the system. The input can be a
float, to set the same charge for all particles, or a list, to define
a unique charge for each particle.</p>
<dl class="simple">
<dt>particles_charge<span class="classifier">float or list</span></dt><dd><p>Charge values to add for the particles in the sbmOpenMM system class.</p>
</dd>
</dl>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.setParticlesHPS">
<span class="sig-name descname"><span class="pre">setParticlesHPS</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">particles_hps</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.setParticlesHPS" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the charge of the particles in the system. The input can be a
float, to set the same charge for all particles, or a list, to define
a unique charge for each particle.</p>
<dl class="simple">
<dt>particles_hps<span class="classifier">float or list</span></dt><dd><p>HPS scale values to add for the particles in the sbmOpenMM system class.</p>
</dd>
</dl>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.setParticlesMasses">
<span class="sig-name descname"><span class="pre">setParticlesMasses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">particles_mass</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.setParticlesMasses" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the masses of the particles in the system. The input can be a
float, to set the same mass for all particles, or a list, to define
a unique mass for each particle.</p>
<dl class="simple">
<dt>particles_mass<span class="classifier">float or list</span></dt><dd><p>Mass(es) values to add for the particles in the sbmOpenMM system class.</p>
</dd>
</dl>
<p>None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="hps.core.system.system.setParticlesRadii">
<span class="sig-name descname"><span class="pre">setParticlesRadii</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">particles_radii</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hps.core.system.system.setParticlesRadii" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the radii of the particles in the system. The input can be a
float, to set the same radius for all particles, or a list, to define
a unique radius for each particle.</p>
<dl class="simple">
<dt>particles_radii<span class="classifier">float or list</span></dt><dd><p>Radii values to add for the particles in the sbmOpenMM system class.</p>
</dd>
</dl>
<p>None</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-hps.core">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-hps.core" title="Permalink to this headline">¶</a></h2>
<p>core package of the sbmOpenMM package that contains the main sbmOpenMM classes.</p>
<p>The sbmOpenMM.core package contains the three sbmOpenMM main classes:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>geometry</p></li>
<li><p>models</p></li>
<li><p>system</p></li>
</ol>
</div></blockquote>
<p>The first class, geometry, contains methods to calculate the geometrical parameters from the input structures. These parameters are used to define the input conformation as the global minimum configuration in the potential energy function. The second class, models, allows to easily set up predefined SBM models, that encompass coarse grained, all atom and multi basin potentials. The third class, system, is the main class that holds all the methods to define, modify and create SBMs to be simulated with OpenMM.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="hps.datasets.html" class="btn btn-neutral float-right" title="hps.datasets package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="hps.html" class="btn btn-neutral float-left" title="hps package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Quyen Vu.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>